slam-libraries/viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -313,34 +340,25 @@ void SLAMServiceImpl::CreateMap() {
             set_start_time = true;
         }
 
-        int num_nodes;
         {
             std::lock_guard<std::mutex> lk(map_builder_mutex);
             auto measurement = map_builder.GetDataFromFile(file);
             if (measurement.ranges.size() > 0) {
                 trajectory_builder->AddSensorData(kRangeSensorId.id,
                                                   measurement);
-                num_nodes = map_builder.map_builder_->pose_graph()
-                                ->GetTrajectoryNodes()
-                                .size();
 
                 auto local_poses = map_builder.GetLocalSlamResultPoses();
                 if (local_poses.size() > 0) {
-                    auto latest_local_pose = local_poses.back();
-                    cartographer::transform::Rigid3d global_pose =
-                        map_builder.GetGlobalPose(trajectory_id,
-                                                  latest_local_pose);
-                    myfile << "global_pose: " << global_pose.DebugString()
-                           << std::endl;
+                    std::lock_guard<std::mutex> lk(pose_mutex);
```

Comment:
[nit] It would be better to avoid taking `pose_mutex` while holding `map_builder_mutex`. Otherwise, we'll always need to impose the ordering that if a codepath is going to take both mutexes, it must take `map_builder_mutex` first. You can do this by copying the pose into a local variable (while holding only `map_builder_mutex`), then after `map_builder_mutex` is released, you can lock `pose_mutex` and copy the local variable into `latest_global_pose`.

If that seems like too much copying, it's okay to leave this as is, but then please add a comment in the header file that any codepath that locks both `map_builder_mutex` and `pose_mutex` must lock `map_builder_mutex` first.

Reply:
+1 on copying rather than imposing an order

Reply:
yeah wasn't sure on doing the double mutex there. I'll move the grpc mutex out.